Fix exp rules for some matrices

sethaxen opened the pull request:
Fixes #595

oxinabox commented on test/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -14,6 +14,11 @@
         end
+        @testset "imbalanced A with no squaring" begin
+            # https://github.com/JuliaDiff/ChainRules.jl/issues/595
+            A = randn(MersenneTwister(8), 3, 3)
```
Maybe using a literal rather than risk the MersenneTwister changing?
(Like it has done in the past)
```suggestion
            A = [
                -0.007623430669065629 -0.567237096385192  0.4419041897734335;
                 2.090838913114862    -1.254084243281689 -0.04145771190198238;
                 2.3397892123412833   -0.6650489083959324 0.6387266010923911
                ]
```

oxinabox commented on test/rulesets/LinearAlgebra/matfun.jl:
```diff
@@ -48,6 +53,11 @@
         end
+        @testset "imbalanced A with no squaring" begin
+            # https://github.com/JuliaDiff/ChainRules.jl/issues/595
+            A = randn(MersenneTwister(8), 3, 3)
```
Maybe using a literal rather than risk the MersenneTwister changing?
(Like it has done in the past)
```suggestion
            A = [
                -0.007623430669065629 -0.567237096385192  0.4419041897734335;
                 2.090838913114862    -1.254084243281689 -0.04145771190198238;
                 2.3397892123412833   -0.6650489083959324 0.6387266010923911
                ]
```

oxinabox reviewed (approved):
nice work. merge and tag when happy

StevenWhitaker commented:
Thanks for working on this @sethaxen. I noticed that there are other random seeds that still fail (I'll post specifics in the #595); do you want to merge this anyway, or wait to resolve the other issues?

sethaxen commented:
> Thanks for working on this @sethaxen. I noticed that there are other random seeds that still fail (I'll post specifics in the #595); do you want to merge this anyway, or wait to resolve the other issues?

Thanks, @StevenWhitaker! It turns out the remaining seeds failed due to truncation error due to not using all of the intermediates. I fixed that also in this PR and added a test that checks 100 random matrices (I typically found one of the first 10 or so would fail before this fix).

Can you verify that this now works for all of your seeds?

StevenWhitaker commented:
Everything looks great on my end!

StevenWhitaker commented:
Out of curiosity, when will this get merged and tagged? @sethaxen

sethaxen commented:
Thanks for the reminder, I think this is good to go, would be good to get one last review from @oxinabox before merging.

StevenWhitaker commented:
Bump

sethaxen commented:
@oxinabox, can you re-review this PR before I merge?

mcabbott reviewed (approved):
I have not worked through the maths here. But it seems to have good tests!
